### PR TITLE
Generate Diff and Patch direct from Pull head

### LINKED ISSUE #10932: [Bug] 500 on download pull-diff/patch file of a merged with deleted HeadBranch
test: https://gitea.com/gitnex/GitNex/pulls/340.diff

Gitea Version: 1.12.0+dev-82-g3723b0647

### PR DESCRIPTION
**Pull request by zeripath**
Fix #10932
Also fix "Empty Diff/Patch File when pull is merged"

Closes #10934

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
New tests also include steps to ensure that the pulls pages (including commits and files) don't produce a 500 even when the head repository is deleted.

File: integrations/git_test.go [hunk 1]
```diff
@@ -419,8 +417,62 @@ func doMergeFork(ctx, baseCtx APITestContext, baseBranch, headBranch string) fun
...
 			assert.NoError(t, err)
 		})
+		t.Run("EnsureCanSeePull", func(t *testing.T) {
+			req := NewRequest(t, "GET", fmt.Sprintf("/%s/%s/pulls/%d", url.PathEscape(baseCtx.Username), url.PathEscape(baseCtx.Reponame), pr.Index))
+			ctx.Session.MakeRequest(t, req, http.StatusOK)
+			req = NewRequest(t, "GET", fmt.Sprintf("/%s/%s/pulls/%d/files", url.PathEscape(baseCtx.Username), url.PathEscape(baseCtx.Reponame), pr.Index))
+			ctx.Session.MakeRequest(t, req, http.StatusOK)
+			req = NewRequest(t, "GET", fmt.Sprintf("/%s/%s/pulls/%d/commits", url.PathEscape(baseCtx.Username), url.PathEscape(baseCtx.Reponame), pr.Index))
+			ctx.Session.MakeRequest(t, req, http.StatusOK)
+		})
```
Comment (bagasme) [inline integrations/git_test.go:427]: What this check doing related to this PR?

Comment (6543) [reply integrations/git_test.go:427]: It check if the permissions are corect interpreted

**Review by 6543 — APPROVED** [review]

File: integrations/git_test.go (hunk 1 above)
Comment (6543) [reply integrations/git_test.go:427]: so if if fails later ... it is not because of some permissions

**Comment by 6543** [comment]
@zeripath this need a backport too

File: integrations/git_test.go (hunk 1 above)
Comment (zeripath) [reply integrations/git_test.go:427]: It's not so much related to this PR as extending the test in general - however, this is likely the first test to ensure that the pulls page and it's files and commits pages work.

We then ensure that they work after the PR is merged, after the head branch is deleted and after the head repository is deleted. 

Probably we should do the test with deleting head branch and repository before merging too.

Comment (zeripath) [reply integrations/git_test.go:427]: But that can be done as a different pr/test.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Please send backport.

**Pull request merged**